### PR TITLE
Adding extra logic for container restart

### PR DESCRIPTION
**Pull request by CharlyF**
### What does this PR do?

When using the command `docker restart <container>` the new container spun up has the same container id.
In order to support short lived containers we introduced a delay to unschedule a check scheduled from the said container.

The issue with the restart is that we remove the check from the cache of known services and wait 5 seconds to unschedule the check. Because we receive an event for a container create and we don't know it, we schedule a check but soon the 5 seconds delay expires and unschedules the newly created check.

Moving the logic to remove the service from the map to the 5 second routine is not enough to solve the issue because when receiving the start event, we would find the container and just end up in [this part of the logic](https://github.com/DataDog/datadog-agent/blob/master/pkg/autodiscovery/listeners/docker.go#L187)

I could see only two solutions:
1/ Given that the methods are stateless, we would need to add logic to know if we are in a restart state instead of a start. This would mean a cache in the listener and then logic to cancel the removal of the check or ensure the subsequent creation of the check after the deletion. 
This felt like too big of a patch, not clean and would add potential leaking point or issues.
2/ Delay the creation of the check in the restart state. As we move the deletion of the service to the delayed routine if we receive the start event within 5 seconds (case of a normal restart) we delay the creation for 5 seconds and we have the check correctly unscheduled as a result of the `die` event and scheduled as a result of the `start` event.
If the restart takes more than 5 seconds, then the service will be removed by the removeservice and as we receive a start event, we will not find the service in the map and thus we follow the regular process.

This is not ideal, but is the least disruptive way I found to fix this issue. 

### Motivation

Customer issue

### Additional Notes

Anything else we should know when reviewing?

### CODE REVIEW

File: pkg/autodiscovery/listeners/docker.go [hunk 1]
```diff
@@ -182,6 +182,11 @@ func (l *DockerListener) processEvent(e *docker.ContainerEvent) {
 	if found {
 		if e.Action == "die" {
 			l.removeService(cID)
+		} else if e.Action == "start" {
```
Comment (clamoriniere) [inline pkg/autodiscovery/listeners/docker.go]: small nit: this `if {} else if {} else{} can be replaced by a "switch case"

**Review by clamoriniere — COMMENTED** [review]
I really like this approach !!! 
just a small nit (that will a potential warning with `golangci-lint`) [ifElseChain check doc](https://github.com/go-critic/go-critic/blob/c77c5c4cc248442409aae64d59105daecca5efcf/docs/overview.md#ifElseChain-ref)

File: pkg/autodiscovery/listeners/docker.go (hunk 1 above)
Comment (CharlyF) [reply pkg/autodiscovery/listeners/docker.go]: yeah true - I just use switch cases when I want to suggest that there can be more use cases I have not considered or to be added in the future. Whereas here we should not have more. If that makes sense

**Review by clamoriniere — APPROVED** [review]
👍

File: pkg/autodiscovery/listeners/docker.go [hunk 2]
```diff
@@ -180,9 +180,15 @@ func (l *DockerListener) processEvent(e *docker.ContainerEvent) {
 	l.m.RUnlock()
 
 	if found {
-		if e.Action == "die" {
+		switch e.Action {
+		case "die":
 			l.removeService(cID)
-		} else {
+		case "start":
+			// Container restarted with the same ID within 5 seconds.
```
Comment (mfpierre) [inline pkg/autodiscovery/listeners/docker.go]: is `docker restart` the only way we could receive this event?

Comment (CharlyF) [reply pkg/autodiscovery/listeners/docker.go]: That I can think of, yes. But even if it's not the only one, this should yield this behavior:
"stop: wait for 5 seconds before removing completely"
"start of the same container within the 5 seconds: wait 5 seconds before starting a new service"

**Review by mfpierre — APPROVED** [review]
:shipit:

**Pull request merged**